### PR TITLE
Add support for Fortran program in the PSyIR (closes #1137)

### LINKED ISSUE #1137: Support a fortran program in the PSyIR
At the moment a Fortran program will produce a code block in the PSyIR when processed using the `fparser2reader` class' `generate_psyir` method. There is an unused option in a `Routine` node (`is_program`) which specifies that the instance is a program. This could be used to translate from an fparser2 ast to PSyIR. Further, the Fortran back-end does not seem to support a Routine's `is_program` option so would also need to be updated.

### PR DESCRIPTION
**Pull request by rupertford**

### CODE REVIEW

**Comment by rupertford** [comment]
I think this is ready for first review from @sergisiso

**Review by sergisiso — CHANGES_REQUESTED** [review]
@rupertford There is just some picky comments but the code looks good and is 100% covered, tests pass also with compilation, no TODOs pending, no documentation is needed and no formatting issues are introduced. Should be ready when the comments are addressed.

File: src/psyclone/psyir/backend/fortran.py [hunk 1]
```diff
@@ -773,9 +776,12 @@ def routine_node(self, node):
             "".format(imports, declarations, exec_statements))
 
         self._depth -= 1
+        name = "subroutine"
+        if node.is_program:
+            name = "program"
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/backend/fortran.py:781]: Updated.

File: src/psyclone/tests/psyir/frontend/fparser2_main_program_handler_test.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+    processor = Fparser2Reader()
+    reader = FortranStringReader(code)
+    parse_tree = parser(reader)
+    program = parse_tree.children[0]
+    psyir = processor._main_program_handler(program, None)
+    # Check the expected PSyIR nodes are being created
+    assert isinstance(psyir, Routine)
+    assert psyir.is_program
+    assert psyir.parent is None
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/psyir/frontend/fparser2_main_program_handler_test.py:93]: Yes this is a good point. Actually this was already the case before this PR as we could also already have `Routine` with `is_program=False` as a top level node. There are two relevant things I would like to say about this from a Fortran code point of view
1) I think the top level for Fortran codes should probably be a FileContainer (or similar name) which is a Container as this maps naturally to a `Program` in the Fortran spec, which can contain one or more modules, programs, subroutines and functions (with a constraint that there can be at most one program). The idea of a File containing Routines (classes, functions, procedures, etc) and being used to scope symbols is also used in other languages, so is pretty generic.
2) Whilst we don't have examples of this type of code at the moment, a Fortran program, subroutine and function can encapsulate other subroutines and functions in the same way that modules do (although I'm not sure about variable scoping in these cases i.e. if there is an equivalent of module scope). Therefore we need a way of capturing this, which I think would naturally lead to integrating the concepts of Container and Routine somehow. One option would be to say that a Routine is a Container, but there are a number of options!

**Comment by rupertford** [comment]
Ready for next review from @sergisiso subject to the CI being happy.

File: src/psyclone/tests/psyir/frontend/fparser2_main_program_handler_test.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sergisiso) [reply src/psyclone/tests/psyir/frontend/fparser2_main_program_handler_test.py:93]: Good points both

**Review by sergisiso — APPROVED** [review]
@rupertford This PR is ready to be merged

**Pull request merged**